server/src/main/java/io/druid/query/ResultLevelCachingQueryRunner.java

```diff
@@ -0,0 +1,298 @@
+            new Function<T, T>()
+            {
+              @Override
+              public T apply(T input)
+              {
+                cacheResultEntry(resultLevelCachePopulator, input);
+                return input;
+              }
+            }
+        ), new SequenceWrapper()
+        {
+          @Override
+          public void after(boolean isDone, Throwable thrown) throws Exception
+          {
+            if (resultLevelCachePopulator != null) {
+              // The resultset identifier and its length is cached along with the resultset
+              resultLevelCachePopulator.populateResults(newResultSetId);
+              log.info("Cache population complete for query %s", query.getId());
+            }
+          }
+        });
+      }
+    } else {
+      return baseRunner.run(
+          queryPlus,
+          responseContext
+      );
+    }
+  }
+
+  private T cacheResultEntry(
+      ResultLevelCachePopulator resultLevelCachePopulator,
+      T resultEntry
+  )
+  {
+    final Function<T, Object> cacheFn = strategy.prepareForResultLevelCache();
+    if (resultLevelCachePopulator != null) {
+      resultLevelCachePopulator.cacheFutures
+          .add(backgroundExecutorService.submit(() -> cacheFn.apply(resultEntry)));
+    }
+    return resultEntry;
+  }
+
+  private byte[] fetchResultsFromResultLevelCache(
+      final String queryCacheKey
+  )
+  {
+    if (useResultCache && queryCacheKey != null) {
+      return cache.get(ResultLevelCacheUtil.computeResultLevelCacheKey(queryCacheKey));
+    }
+    return null;
+  }
+
+  private String extractEtagFromResults(
+      final byte[] cachedResult
+  )
+  {
+    if (cachedResult == null) {
+      return null;
+    }
+    log.info("Fetching result level cache identifier for query: %s", query.getId());
+    int etagLength = ByteBuffer.wrap(cachedResult, 0, Integer.BYTES).getInt();
+    return StringUtils.fromUtf8(Arrays.copyOfRange(cachedResult, Integer.BYTES, etagLength + Integer.BYTES));
+  }
+
+  private Sequence<T> deserializeResults(
+      final byte[] cachedResult, CacheStrategy strategy, String resultSetId
+  )
+  {
+    if (cachedResult == null) {
+      return null;
+    }
+    final Function<Object, T> pullFromCacheFunction = strategy.pullFromResultLevelCache();
+    final TypeReference<Object> cacheObjectClazz = strategy.getCacheObjectClazz();
+    //Skip the resultsetID and its length bytes
+    byte[] prunedCacheData = Arrays.copyOfRange(
+        cachedResult,
+        Integer.BYTES + resultSetId.length(),
+        cachedResult.length
+    );
+    Sequence<T> cachedSequence = Sequences.simple(() -> {
+      try {
+        if (cachedResult.length == 0) {
+          return Iterators.emptyIterator();
+        }
+
+        return objectMapper.readValues(
+            objectMapper.getFactory().createParser(prunedCacheData),
+            cacheObjectClazz
+        );
+      }
+      catch (IOException e) {
+        throw new RuntimeException(e);
```

Comment:
add some error message too such as what failed, what was query id etc etc. and , use "RE" class.